File: 2020/Python/07-HandyHaversack/HandyHaversack.py
#!/usr/bin/env python3

# Handy Haversack
# https://adventofcode.com/2020/day/7

def read_file(name):
    file = open(name)
    return list(file.readlines())

def parse_rule(rule):
    outer_bag, allowable_bag_string = rule.split(" bags contain ")
    allowable_bag_components = allowable_bag_string.split(", ")
    allowable_bags = [ parse_bag(bag) for bag in allowable_bag_components if not bag.startswith("no") ]
    return (outer_bag, allowable_bags)

def parse_bag(allowable_bag):
    components = allowable_bag.split(" ")
    count = int(components[0])
    bag = "{0} {1}".format(components[1], components[2])
    return (count, bag)

# Returns bags that can ultimately hold the target bag based on the given rules.
def bags(rules, target):
    result = set()
    bags_to_try = [ target ]
    while bags_to_try:
        bag = bags_to_try.pop()
        if bag in result:
            continue
        result.add(bag)
        next_bags = bags_that_directly_hold(rules, bag)
        bags_to_try.extend(next_bags)
    result.remove(target)
    return list(result)

# Returns bags that can directly hold the target bag based on the given rules.
def bags_that_directly_hold(rules, target):
    return [ bag for bag, allowed_bags in rules.items() if contains(allowed_bags, target) ]

def contains(bags, target):
    return any([ bag for count, bag in bags if bag == target ])

# Returns the sum of all of the bags that are held by the given bag.
def bag_count(rules, bag):
    return sum([ count * (1 + bag_count(rules, inner_bags)) for count, inner_bags in rules[bag] ])

rule_strings = read_file("07-input.txt")
rule_tuples = [ parse_rule(rule) for rule in rule_strings ]
rules = { bag: allowable_bags for bag, allowable_bags in rule_tuples }
bags_that_hold_shiny_gold = bags(rules, "shiny gold")
print("There are {0} bags that can hold shiny gold bags".format(len(bags_that_hold_shiny_gold)))

count = bag_count(rules, "shiny gold")
print("Shiny gold bags hold {0} other bags".format(count))

File: 2020/Swift/07-HandyHaversack/HandyHaversack.swift
#!/usr/bin/swift

// Handy Haversack
// https://adventofcode.com/2020/day/7

import Foundation

func readFile(named name: String) -> [String] {
    let currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    let fileURL = URL(fileURLWithPath: name + ".txt", relativeTo: currentDirectoryURL)
    guard let content = try? String(contentsOf: fileURL, encoding: String.Encoding.utf8) else {
        print("Unable to read input file \(name)")
        print("Current directory: \(currentDirectoryURL)")
        return []
    }
    return content.components(separatedBy: .newlines)
}

func parse(rule: String) -> (String, [(Int, String)])? {
    let ruleComponents = rule.components(separatedBy: " bags contain ")
    guard ruleComponents.count == 2 else { return nil }

    let outerBag = ruleComponents[0]
    let allowableBagsComponents = ruleComponents[1].components(separatedBy: ", ")
    let allowableBags = allowableBagsComponents.compactMap { parse(allowableBag: $0) }
    return (outerBag, allowableBags)
}

func parse(allowableBag: String) -> (Int, String)? {
    let components = allowableBag.components(separatedBy: " ")
    guard components.count == 4 else { return nil }

    guard let count = Int(components[0]) else { return nil }
    let bag = components[1] + " " + components[2]
    return (count, bag)
}

// The list of bags that directly hold the given bag.
func bags(from rules: [String : [(Int, String)]], thatCanDirectlyHold bag: String) -> [String] {
    rules.keys.filter { key in
        guard let allowedBags = rules[key] else { return false }
        return allowedBags.contains { $0.1 == bag }
    }
}

// The list of bags that contain bags that ultimately hold the given bag.
func bags(from rules: [String : [(Int, String)]], thatCanHold bag: String) -> [String] {
    var result = Set<String>()
    var triedBags = Set<String>()
    var bagsToTry = [bag]
    while !bagsToTry.isEmpty {
        let nextBag = bagsToTry.removeFirst()
        guard !triedBags.contains(nextBag) else { continue }
        triedBags.insert(nextBag)

        let nextBags = bags(from: rules, thatCanDirectlyHold: nextBag)
        result.formUnion(Set(nextBags))
        bagsToTry.append(contentsOf: nextBags)
    }
    return Array(result)
}

// The sum of all of the bags that are held by the given bag.
func bagCount(from rules: [String : [(Int, String)]], for bag: String) -> Int {
    guard let innerBags = rules[bag] else { return 0 }
    return innerBags.map { $0.0 * (1 + bagCount(from: rules, for: $0.1)) }.reduce(0, +)
}

let ruleStrings = readFile(named: "07-input")
let rules = ruleStrings.compactMap { parse(rule: $0) }.reduce(into: [:]) { $0[$1.0] = $1.1 }
let bagsThatHoldShinyGold = bags(from: rules, thatCanHold: "shiny gold")
print("There are \(bagsThatHoldShinyGold.count) bags that can hold shiny gold bags")

let count = bagCount(from: rules, for: "shiny gold")
print("Shiny gold bags hold \(count) other bags")
